tee-worker/litentry/core/data-providers/src/nodereal_jsonrpc.rs

```diff
@@ -0,0 +1,343 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{build_client, Error, HttpError, GLOBAL_DATA_PROVIDER_CONFIG};
+use http::header::CONNECTION;
+use http_req::response::Headers;
+use itc_rest_client::{
+	http_client::{DefaultSend, HttpClient},
+	rest_client::RestClient,
+	RestPath, RestPost,
+};
+use log::debug;
+use serde::{Deserialize, Serialize};
+use std::{
+	format, str,
+	string::{String, ToString},
+	thread, time, vec,
+	vec::Vec,
+};
+
+pub enum NoderealChain {
+	// BNB Smart Chain
+	Bsc,
+	// opBNB
+	Opbnb,
+	// Ethereum
+	Eth,
+	// ETH Beacon Chain
+	Eth2Beacon,
+	// Aptos
+	Aptos,
+	// Optimism
+	Opt,
+	// Polygon
+	Polygon,
+}
+
+impl NoderealChain {
+	pub fn to_string(&self) -> &'static str {
+		match self {
+			NoderealChain::Bsc => "bsc",
+			NoderealChain::Opbnb => "opbnb",
+			NoderealChain::Eth => "eth",
+			NoderealChain::Eth2Beacon => "eth2-beacon",
+			NoderealChain::Aptos => "aptos",
+			NoderealChain::Opt => "opt",
+			NoderealChain::Polygon => "polygon",
+		}
+	}
+}
+
+#[derive(Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ReqPath {
+	path: String,
+}
+
+impl ReqPath {
+	pub fn new(api_key: &str) -> Self {
+		Self { path: "v1/".to_string() + api_key }
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReqBody<P>
+where
+	P: Serialize,
+{
+	pub id: i64,
+	pub jsonrpc: String,
+	pub method: String,
+	pub params: P,
+}
+
+impl<P> ReqBody<P>
+where
+	P: Serialize,
+{
+	fn new(jsonrpc: &str, method: &str, params: P) -> Self {
+		ReqBody { id: 1, jsonrpc: jsonrpc.to_string(), method: method.to_string(), params }
+	}
+}
+
+impl<P> RestPath<ReqPath> for ReqBody<P>
+where
+	P: Serialize,
+{
+	fn get_path(req: ReqPath) -> core::result::Result<String, HttpError> {
+		Ok(req.path)
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct RespBody {
+	pub id: i64,
+	pub jsonrpc: String,
+	pub result: serde_json::Value,
+}
+
+pub struct NoderealJsonrpcClient {
+	api_key: String,
+	api_retry_delay: u64,
+	api_retry_times: u16,
+	client: RestClient<HttpClient<DefaultSend>>,
+}
+
+impl NoderealJsonrpcClient {
+	pub fn new(chain: NoderealChain) -> Self {
+		let api_key = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_key.clone();
+		let api_retry_delay = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_retry_delay;
+		let api_retry_times = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_retry_times;
+		let api_url = GLOBAL_DATA_PROVIDER_CONFIG
+			.write()
+			.unwrap()
+			.nodereal_api_chain_network_url
+			.clone();
+		let base_url =
+			api_url.replace("{chain}", chain.to_string()).replace("{network}", "mainnet");
```

Comment:
If we always use "mainnet" why replace it ?

Reply:
Actually don't get where the "mainnet" come from ...